<<<<<<< HEAD
---
title: "R Notebook"
output: html_notebook
---



```{r}
library(ggplot2)

library(scico)

library(tidyverse)

library(rjson)

today = Sys.Date()

# global ggplot2 theme
theme_set(
  theme_minimal() +
    theme(
      text = element_text(
        size = 16,
        family = "Helvetica",
        colour = "black"
      ),
      axis.text = element_text(colour = "black"),
      plot.title = element_text(hjust = 0.5),
      strip.text = element_text(colour = "black")
    )
)
```

## flower pool-seq: do frequency estimates correlate with coverage?

```{r}

freqk_files <- list.files(path = "/global/scratch/users/milesroberts/moi_lab_projects/grenenet-phase2/workflow/freqk_call_results", pattern = ".txt", full.names = T)

cov_files <- list.files(path = "/global/scratch/users/milesroberts/moi_lab_projects/grenenet-phase2/workflow/fastp_results", pattern = ".json", full.names = T)

extract_num_bases = function(x){
  json_ob <- fromJSON(file = x)
  json_ob$read2_after_filtering$total_reads
}

extract_sample_id = function(x){
  gsub("/global/scratch/users/milesroberts/moi_lab_projects/grenenet-phase2/workflow/fastp_results/", "", gsub("_R1R2.json", "", x))
}

extract_freqk = function(x){
  read.table(x, sep = "|")$V1[1]
}

extract_freqk_id = function(x){
  gsub(".*/", "", gsub(".txt", "", x))
}

coverages <- data.frame(
  ID = unlist(lapply(cov_files, extract_sample_id)),
  num_bases = unlist(lapply(cov_files, extract_num_bases))
)

freqs <- data.frame(
  ID = unlist(lapply(freqk_files, extract_freqk_id)),
  freq = unlist(lapply(freqk_files, extract_freqk))
)

freq_cov <- merge(freqs, coverages, by = "ID")

# distribution of REF frequencies across samples
ggplot(freq_cov, aes(x = freq)) +
  geom_density()

ggplot(freq_cov[(freq_cov$freq > 0),], aes(x = freq)) +
  geom_density()

# do frequencies correlate with coverage
# ideally, no!
sub_freq_cov <- subset(freq_cov, (freq > 0) & (freq < 1))

ggplot(sub_freq_cov, aes(x = num_bases, y = freq)) +
  geom_point() +
  scale_x_log10() +
  labs(x = "Number of bases (post-fastp)", y = "REF frequency", title = paste("\u03c1 =", signif(cor.test(sub_freq_cov$num_bases, sub_freq_cov$freq, method = "spearman")$estimate, 2)))

ggplot(freq_cov, aes(x = num_bases, y = freq)) +
  geom_point() +
  scale_x_log10() +
  labs(x = "Number of bases (post-fastp)", y = "REF frequency", title = paste("r =", signif(cor.test(freq_cov$num_bases, freq_cov$freq, method = "pearson")$estimate, 2)))

ggsave(paste("../results/", today, "/coverage_vs_freqk_estimates.pdf", sep = ""),height = 5, width = 5)
```

## seed mix pool-seq: do freqk estimates correlate with known SNP frequencies?

```{r}
# freqk estimates
freqk_calls <- read.table("/global/scratch/users/milesroberts/moi_lab_projects/grenenet-phase2/resources/calls.txt", sep = "|", header = F, col.names = c("RF_freqk", "AF_freqk"))

freqk_index <- read.table("/global/scratch/users/milesroberts/moi_lab_projects/grenenet-phase2/resources/ref_index_greneNet.txt", sep = ",", header = F)

freqk_calls <- cbind(freqk_index, freqk_calls)

freqk_calls$V3 = freqk_calls$V3 + 2

# greneNet vcf
real_calls <- read.table("/global/scratch/users/milesroberts/moi_lab_projects/grenenet-phase2/resources/greneNet_allele_counts.tsv", sep = "\t", header = F, col.names = c("CHROM", "POS", "REF", "ALT", "AN", "AC"))

# calculate reference allele frequency
real_calls$RF_real = 1 - real_calls$AC/real_calls$AN

# merge
names(freqk_calls)[2:3] = c("CHROM", "POS")

real_vs_freqk <- merge(real_calls, freqk_calls, by = c("CHROM", "POS"))

# plot
rho_estimate = signif(cor.test(real_vs_freqk$RF_real, real_vs_freqk$RF_freqk, method = "spearman")$estimate, 3)

ggplot(real_vs_freqk, aes(x = RF_real, y = RF_freqk)) +
  geom_bin2d() +
  #geom_smooth() +
  geom_abline(slope = 1, intercept = 0, color = "white", lwd = 2) +
  #scale_fill_continuous(trans = "log10") +
  scale_fill_scico(palette = "bilbao", trans = "log10") +
  theme_minimal() +
  theme(text = element_text(size = 16, family = "Helvetica", colour = "black"),
        axis.text.x = element_text(color = "black"),
        axis.text.y = element_text(color = "black"), 
        plot.title = element_text(hjust = 0.5)) +
  labs(x = "REF frequency in greneNet VCF", y = "REF frequency from\nk-mer counts", fill = "log10(count)", title = paste("\u03C1 = ", rho_estimate, sep = ""))

ggsave(paste("../results/", today, "/bin2d_greneNet_vs_freqk.pdf", sep = ""), height = 5, width = 7, device = cairo_pdf)
```

## Frequency trends in cam5indel
```{r}
names(freq_cov)[1] <- "sampleid"

# merge sample information with frequency estimates
freq_samples <- merge(freq_cov, samples_data, by = "sampleid")

freq_summary <- freq_samples %>%
  dplyr::group_by(site, generation) %>%
  dplyr::summarise(mean_freq = mean(freq, na.rm = T))

ggplot(freq_summary, aes(x = generation, y = mean_freq)) +
  geom_point() +
  geom_line() +
  facet_wrap(~site, ncol = 4, nrow = 8)

ggsave(paste("../results/", today, "/temporal_freq_cam5indel.pdf", sep = ""), height = 9, width = 5, device = cairo_pdf)
```

=======
---
title: "R Notebook"
output: html_notebook
---


## set-up

```{r}
rm(list = ls())

library(ggplot2)

library(scico)

library(reshape2)

library(dplyr)

library(purrr)

today <- Sys.Date()

# global ggplot2 theme
theme_set(
  theme_minimal() +
    theme(
      text = element_text(
        size = 16,
        family = "Helvetica",
        colour = "black"
      ),
      axis.text = element_text(colour = "black"),
      plot.title = element_text(hjust = 0.5),
      strip.text = element_text(colour = "black")
    )
)
```

## functions

```{r}

plot_manhattan_by_site <- function(data, site, today, height, width){

print(site)
  
ggplot(subset(data, SITE == site), aes(x = POS, y = -log10(pvalue))) +
  geom_point(shape = 16) +
  facet_wrap(~CHROM, ncol = 5, labeller = "label_both", scales = "free_x") +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  labs(x = "Position", y = "-log10(p-value)")
  
ggsave(paste("../results/", today, "/manhattan_feder_", site, ".pdf", sep = ""), height = height, width = width)

}

```

## Manhattan plots of sweep statistics in founders

```{r}

sweeps <- read.table("../data/sweeps.txt", header = T, sep = "\t")

sweeps <- melt(sweeps, id = c("Contig", "Var_Start", "Var_Stop", "Pos_Start", "Pos_Stop"))

sweeps <- sweeps[!(sweeps$variable %in% c("Theta_W", "Theta_Pi")),]

# manhattan plots of 
sweeps$variable <- gsub("_", " ", sweeps$variable)

sweeps$Contig <- as.factor(sweeps$Contig)

sweeps$Contig <- revalue(sweeps$Contig, c("0" = "1", "1" = "2", "2" = "3", "3" = "4", "4" = "5"))

ggplot(sweeps, aes(x = (Pos_Start + Pos_Stop)/2, y = value)) +
  geom_point(shape = 16) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  facet_wrap(variable~Contig, ncol = 5, nrow = 5, scales = "free") +
  labs(x = "Position", y = "Statistic")

ggsave(paste("../results/", today, "/sweep_stats.pdf", sep = ""), height = 11, width = 13)
```

## seed mix: compare freqk estimated allele frequencies to vcf snp calls

```{r}

# freqk estimates
freqk_calls <- read.table("/global/scratch/users/milesroberts/moi_lab_projects/grenenet-phase2/resources/calls.txt", sep = "|", header = F, col.names = c("RF_freqk", "AF_freqk"))

freqk_index <- read.table("/global/scratch/users/milesroberts/moi_lab_projects/grenenet-phase2/resources/ref_index_greneNet.txt", sep = ",", header = F)

freqk_calls <- cbind(freqk_index, freqk_calls)

freqk_calls$V3 = freqk_calls$V3 + 2

# greneNet vcf
real_calls <- read.table("/global/scratch/users/milesroberts/moi_lab_projects/grenenet-phase2/resources/greneNet_allele_counts.tsv", sep = "\t", header = F, col.names = c("CHROM", "POS", "REF", "ALT", "AN", "AC"))

# calculate reference allele frequency
real_calls$RF_real = 1 - real_calls$AC/real_calls$AN

# merge
names(freqk_calls)[2:3] = c("CHROM", "POS")

real_vs_freqk <- merge(real_calls, freqk_calls, by = c("CHROM", "POS"))


# plot
rho_estimate = signif(cor.test(real_vs_freqk$RF_real, real_vs_freqk$RF_freqk, method = "spearman")$estimate, 3)

ggplot(real_vs_freqk, aes(x = RF_real, y = RF_freqk)) +
  geom_bin2d() +
  #geom_smooth() +
  geom_abline(slope = 1, intercept = 0, color = "white", lwd = 2) +
  #scale_fill_continuous(trans = "log10") +
  scale_fill_scico(palette = "bilbao", trans = "log10") +
  theme_minimal() +
  theme(text = element_text(size = 16, family = "Helvetica", colour = "black"),
        axis.text.x = element_text(color = "black"),
        axis.text.y = element_text(color = "black"), 
        plot.title = element_text(hjust = 0.5)) +
  labs(x = "REF frequency in greneNet VCF", y = "REF frequency from\nk-mer counts", fill = "log10(count)", title = paste("\u03C1 = ", rho_estimate, sep = ""))

ggsave(paste("../results/", today, "/bin2d_greneNet_vs_freqk.pdf", sep = ""), height = 5, width = 7, device = cairo_pdf)
```

## find allele frequency data

```{r}

gen0 <- read.table("../data/average_seedmix_p0_LDpruned.txt", sep = "\t")

hapfire <- read.table("../data/merged_hapFIRE_allele_frequency_LDpruned.txt", sep = "\t", header = T)

gen1 <- hapfire[,grepl("_1_", names(hapfire))] 

gen2 <- hapfire[,grepl("_2_", names(hapfire))] 

gen3 <- hapfire[,grepl("_3_", names(hapfire))] 

#
shared_gen1 <- (gsub("_1_","_",names(gen1)) %in% gsub("_2_","_",names(gen2))) & (gsub("_1_","_",names(gen1)) %in% gsub("_3_","_",names(gen3)))

shared_gen2 <- (gsub("_2_","_",names(gen2)) %in% gsub("_3_","_",names(gen3))) & (gsub("_2_","_",names(gen2)) %in% gsub("_1_","_",names(gen1)))

shared_gen3 <- (gsub("_3_","_",names(gen3)) %in% gsub("_2_","_",names(gen2))) & (gsub("_3_","_",names(gen3)) %in% gsub("_1_","_",names(gen1)))

gen1_sub <- gen1[,shared_gen1]

gen2_sub <- gen2[,shared_gen2]

gen3_sub <- gen3[,shared_gen3]

# calculate statistic
y1 <- (gen1_sub - gen0$V3)/sqrt(2*gen0$V3*(1-gen0$V3)*(1-0))
y1$gen <- 1
y1$CHROM <- gen0$V1
y1$POS <- gen0$V2
y1 <- melt(y1, id = c("gen", "CHROM", "POS"))


y2 <- (gen2_sub - gen1_sub)/sqrt(2*gen1_sub*(1-gen1_sub)*(2-1))
y2$gen <- 2
y2$CHROM <- gen0$V1
y2$POS <- gen0$V2
y2 <- melt(y2, id = c("gen", "CHROM", "POS"))

y3 <- (gen3_sub - gen2_sub)/sqrt(2*gen2_sub*(1-gen2_sub)*(2-1))
y3$gen <- 3
y3$CHROM <- gen0$V1
y3$POS <- gen0$V2
y3 <- melt(y3, id = c("gen", "CHROM", "POS"))

yvalues <- rbind(y1, y2, y3)
yvalues$SITE <- gsub("X", "", gsub("_.*", "", yvalues$variable))
yvalues$REP <- gsub(".*_", "", yvalues$variable)

t_test_stat <- yvalues %>% group_by(CHROM, POS, SITE) %>% summarize(ybar = mean(value, na.rm = T), s2 = var(value, na.rm = T), n = n())

t_test_stat$tvalue <- t_test_stat$ybar/sqrt(t_test_stat$s2/t_test_stat$n)

t_test_stat$pvalue <- dt(t_test_stat$tvalue, df = t_test_stat$n)

# Manhattan plot
map(unique(t_test_stat$SITE), ~plot_manhattan_by_site(data = t_test_stat, site = .x, today, height = 4, width = 10))

#load("../data/allele_allyears_frequencies_long_raw_climate.rda")
# calculate mean and average

# ybar <- (y1 + y2 + y3)/3
# 
# s2 <- (1/(3-1))*( (y1 - ybar)^2 + (y2 - ybar)^2 + (y3 - ybar)^2 )
# 
# t_test_stat <- ybar/sqrt(s2/3)
# 
# # 
# t_test_stat$CHROM <- gen0$V1
# 
# t_test_stat$POS <- gen0$V2
# 
# t_test_stat <- melt(t_test_stat, id = c("CHROM", "POS"))
# 
# t_test_stat$SITE <- gsub("X", "", gsub("_.*", "", t_test_stat$variable))
# 
# t_test_stat$REP <- gsub(".*_", "", t_test_stat$variable)
# 
# # average across reps of each pos at each site
# mean_t_stat <- t_test_stat %>% 
#   group_by(CHROM, POS, SITE) %>% 
#   summarize(MEANT = mean(value, na.rm = T), SDT = sd(value, na.rm = T))
# 
# # calculate p-values
# #
# mean_t_stat$PVALUE <- dt(mean_t_stat$MEANT, df = 2)
# 
# # adjust p-values
# mean_t_stat$ADJPVALUE <- p.adjust(mean_t_stat$PVALUE, method = "bonferroni")
# 
# # plot distribution of stats across positions and sites
# ggplot(mean_t_stat, aes(x = MEANT)) +
#   geom_density()
#   xlim(c(-10, 10))
#   
# # Manhattan plot
# map(unique(mean_t_stat$SITE), ~manhattan_by_site(data = mean_t_stat, site = .x, today, height = 4, width = 10))

#
#summary(unlist(as.vector(t_test_stat)))

```

>>>>>>> 0c13326536bc926a36e573bb65d6c2d03692229a
