---
title: "R Notebook"
output: html_notebook
---


## set-up

```{r}
library(ggplot2)

library(scico)

library(reshape2)

library(plyr)

today <- Sys.Date()

# global ggplot2 theme
theme_set(
  theme_minimal() +
    theme(
      text = element_text(
        size = 16,
        family = "Helvetica",
        colour = "black"
      ),
      axis.text = element_text(colour = "black"),
      plot.title = element_text(hjust = 0.5),
      strip.text = element_text(colour = "black")
    )
)
```

## Manhattan plots of sweep statistics in founders

```{r}

sweeps <- read.table("../data/sweeps.txt", header = T, sep = "\t")

sweeps <- melt(sweeps, id = c("Contig", "Var_Start", "Var_Stop", "Pos_Start", "Pos_Stop"))

sweeps <- sweeps[!(sweeps$variable %in% c("Theta_W", "Theta_Pi")),]

# manhattan plots of 
sweeps$variable <- gsub("_", " ", sweeps$variable)

sweeps$Contig <- as.factor(sweeps$Contig)

sweeps$Contig <- revalue(sweeps$Contig, c("0" = "1", "1" = "2", "2" = "3", "3" = "4", "4" = "5"))

ggplot(sweeps, aes(x = (Pos_Start + Pos_Stop)/2, y = value)) +
  geom_point(shape = 16) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  facet_wrap(variable~Contig, ncol = 5, nrow = 5, scales = "free") +
  labs(x = "Position", y = "Statistic")

ggsave(paste("../results/", today, "/sweep_stats.pdf", sep = ""), height = 11, width = 13)
```

## seed mix: compare freqk estimated allele frequencies to vcf snp calls

```{r}

# freqk estimates
freqk_calls <- read.table("/global/scratch/users/milesroberts/moi_lab_projects/grenenet-phase2/resources/calls.txt", sep = "|", header = F, col.names = c("RF_freqk", "AF_freqk"))

freqk_index <- read.table("/global/scratch/users/milesroberts/moi_lab_projects/grenenet-phase2/resources/ref_index_greneNet.txt", sep = ",", header = F)

freqk_calls <- cbind(freqk_index, freqk_calls)

freqk_calls$V3 = freqk_calls$V3 + 2

# greneNet vcf
real_calls <- read.table("/global/scratch/users/milesroberts/moi_lab_projects/grenenet-phase2/resources/greneNet_allele_counts.tsv", sep = "\t", header = F, col.names = c("CHROM", "POS", "REF", "ALT", "AN", "AC"))

# calculate reference allele frequency
real_calls$RF_real = 1 - real_calls$AC/real_calls$AN

# merge
names(freqk_calls)[2:3] = c("CHROM", "POS")

real_vs_freqk <- merge(real_calls, freqk_calls, by = c("CHROM", "POS"))


# plot
rho_estimate = signif(cor.test(real_vs_freqk$RF_real, real_vs_freqk$RF_freqk, method = "spearman")$estimate, 3)

ggplot(real_vs_freqk, aes(x = RF_real, y = RF_freqk)) +
  geom_bin2d() +
  #geom_smooth() +
  geom_abline(slope = 1, intercept = 0, color = "white", lwd = 2) +
  #scale_fill_continuous(trans = "log10") +
  scale_fill_scico(palette = "bilbao", trans = "log10") +
  theme_minimal() +
  theme(text = element_text(size = 16, family = "Helvetica", colour = "black"),
        axis.text.x = element_text(color = "black"),
        axis.text.y = element_text(color = "black"), 
        plot.title = element_text(hjust = 0.5)) +
  labs(x = "REF frequency in greneNet VCF", y = "REF frequency from\nk-mer counts", fill = "log10(count)", title = paste("\u03C1 = ", rho_estimate, sep = ""))

ggsave(paste("../results/", today, "/bin2d_greneNet_vs_freqk.pdf", sep = ""), height = 5, width = 7, device = cairo_pdf)
```



